core/app/models/spree/taxon.rb

```diff
@@ -17,10 +17,11 @@ class Taxon < Spree::Base
     before_update :set_permalink
     after_update :update_child_permalinks, if: :saved_change_to_permalink?
 
-    validates :name, presence: true
+    validates :name, presence: true, uniqueness: { scope: :parent_id, message: :must_be_unique_under_same_parent }
     validates :meta_keywords, length: { maximum: 255 }
     validates :meta_description, length: { maximum: 255 }
     validates :meta_title, length: { maximum: 255 }
+    validates :taxonomy_id, uniqueness: { message: :can_have_only_one_root }, if: :root?
```

Comment:
Isn't it more `:root_must_be_unique`? I mean, the error is not that the taxon has many roots, but that the taxonomy has already been taken as a parent.

Regardless, should it be something that users can change? As far as I can see, even if it's [permitted by the controller](https://github.com/nebulab/solidus/blob/5d119ba3d1e05c508a6e991bdc5649174a6096c7/core/lib/spree/permitted_attributes.rb#L121), that's not part of the backend's UI:

**Creation**
![screenshot-localhost_3000-2023 02 23-05_44_18](https://user-images.githubusercontent.com/52650/220823426-b9a935aa-8a00-4cc0-a234-6d4ac169051f.png)

**Update**
![screenshot-localhost_3000-2023 02 23-05_44_49](https://user-images.githubusercontent.com/52650/220823427-0299a061-75d4-4544-8b41-bfcff5bfde46.png)

However, it can be modified via API both in [creation](https://solidus.stoplight.io/docs/solidus/2e0e5c45af59c-create-taxonomy-taxon) and [update](https://solidus.stoplight.io/docs/solidus/18c69a84a5e88-update-taxonomy-taxon).

Given that, it makes sense to add the validation now but should we deprecate the `:taxonomy_id` attribute from the list of modifiable fields in the API?

Reply:
Hmm, I disagree. I think the error and validation is about not allowing multiple roots under one Taxonomy. Parents are not so relevant, as parents are just describing how the Taxons are nested under the Taxonomy. The only relevancy is that root Taxons have no parent.

I am not 100% set on `root_must_be_unique`. It's half correct, as we are not testing only the uniqueness of `taxonomy_id`. We are testing that it's unique when `parent_id` is `nil`, or that all roots have a unique Taxonomy. The following Taxons would be considered "unique" but are not valid as they are both trying to be roots for the same Taxonomy:
- `taxonomy_id: 1, parent_id: nil, name: 'Categories', permalink: 'Categories'`
- `taxonomy_id: 1, parent_id: nil, name: 'Brands', permalink: 'Brands'`.

But this is just semantics over a message name, so I am open to other suggestions.

---

What part would you consider to be changed by the user? I don't think we should let the validation be optional, as Taxonomies assume they only have one root:
https://github.com/solidusio/solidus/blob/2b79f72aa53f5caa850c587888fff46c1c91f7b7/core/app/models/spree/taxonomy.rb#L10

Note, in admin you cannot make attempt to make multiple roots for one Taxonomy. You also can't modify the `taxonomy_id` in the update page. `parent_id` still needs to be permissable as you can move a Taxon's nesting.

Reply:
> Hmm, I disagree. I think the error and validation is about not allowing multiple roots under one Taxonomy. Parents are not so relevant, as parents are just describing how the Taxons are nested under the Taxonomy. The only relevancy is that root Taxons have no parent.

Oh, I see. I was confusing root with parent. The problem is that it's a message that refers to another model, Taxonomy, and not to the Taxon one, so it reads weird. But I can't think of a better name now 🤷 

> What part would you consider to be changed by the user? I don't think we should let the validation be optional, as Taxonomies assume they only have one root:

I mean that the creation of the root taxon is something automatic (unless I'm wrong), so users shouldn't deal with it. But that would be part of another PR.

Reply:
Fair point @waiting-for-dev. I wanted to note, making root Taxons is something users should generally don't have to worry about. When you make a Taxonomy, it automatically makes the root Taxon:
https://github.com/solidusio/solidus/blob/5d119ba3d1e05c508a6e991bdc5649174a6096c7/core/app/models/spree/taxonomy.rb#L27

Though, I think it would still be beneficial to have validation to ensure multiple Taxons under the same Taxonomy don't have the `parent_id` as `nil`. Which would mess up the finding of the root. This can happen easily if you make Taxons via code and forget to set the `parent_id`.

Regarding the message, what do you think about keeping the `can_have_only_one_root` name unless we think of something better and changing the message to: "must be unique if parent is set to nil (Taxonomy can only have one root Taxon)"?

Reply:
Yeah, @RyanofWoods, I'm good with the name 🙂

Reply:
But did you want to keep the message as is or? 🤔

Reply:
Oh, sorry, yeah, I'm good 👍